entities/sync/sync.go

```diff
@@ -9,104 +9,173 @@
+	// If we're at capacity, try to remove the oldest entry
+	for kl.count.Load() >= kl.maxSize {
+		var oldestKey interface{}
+		var oldestTime int64 = now + 1 // Initialize to future time
 
-	iLock = iLocks.(*sync.RWMutex)
-	iLock.Lock()
+		// Find the oldest entry
+		kl.locks.Range(func(k, v interface{}) bool {
+			entry := v.(*lockEntry)
+			lastUsed := entry.lastUsed.Load()
+			if lastUsed < oldestTime && !isLocked(&entry.lock) {
+				oldestTime = lastUsed
+				oldestKey = k
+			}
+			return true
+		})
+
+		// Break if we couldn't find any unlocked entries
+		if oldestKey == nil {
+			break
+		}
+
+		// Try to remove the oldest entry
+		kl.locks.Delete(oldestKey)
+		kl.count.Add(-1)
+	}
+
+	// Create and store new entry
+	entry := &lockEntry{}
+	entry.lastUsed.Store(now)
+
+	actual, loaded := kl.locks.LoadOrStore(key, entry)
+	if loaded {
+		entry = actual.(*lockEntry)
+		entry.lastUsed.Store(now)
+	} else {
+		kl.count.Add(1)
+	}
+	entry.lock.Lock()
 }
 
-// Unlock it unlocks a specific item by it's ID
-func (s *KeyRWLocker) Unlock(ID string) {
-	iLocks, _ := s.m.Load(ID)
-	iLock := iLocks.(*sync.RWMutex)
-	iLock.Unlock()
+// Unlock releases the lock for the given key.
+// It must be called exactly once for each successful Lock call.
+// Calling Unlock on an unlocked key or a key that doesn't exist
+// may result in undefined behavior.
+func (kl *KeyLocker) Unlock(key string) {
+	value, _ := kl.locks.Load(key)
+	entry := value.(*lockEntry)
+	entry.lock.Unlock()
 }
 
-// RLock it rlocks a specific bucket by it's ID
-// to hold ant concurrent access to that specific item
+// Cleanup removes all locks that haven't been used for longer than maxAge.
+// This method is called automatically by the periodic cleanup goroutine,
+// but can also be called manually if immediate cleanup is desired.
 //
-//	do not forget calling RUnlock() after rlocking it.
-func (s *KeyRWLocker) RLock(ID string) {
-	iLock := &sync.RWMutex{}
-	iLocks, _ := s.m.LoadOrStore(ID, iLock)
+// It is safe to call this method while other goroutines are using the KeyLocker.
+// Only unused and expired locks will be removed.
+func (kl *KeyLocker) Cleanup() {
+	threshold := time.Now().Add(-kl.maxAge).UnixNano()
+	removed := int32(0)
 
-	iLock = iLocks.(*sync.RWMutex)
-	iLock.RLock()
+	kl.locks.Range(func(key, value interface{}) bool {
+		entry := value.(*lockEntry)
+		if entry.lastUsed.Load() < threshold && !isLocked(&entry.lock) {
+			kl.locks.Delete(key)
+			removed++
+		}
+		return true
+	})
+
+	if removed > 0 {
+		kl.count.Add(-removed)
+	}
+}
+
+func (kl *KeyLocker) periodicCleanup(interval time.Duration) {
+	ticker := time.NewTicker(interval)
+	for range ticker.C {
+		kl.Cleanup()
+	}
 }
 
-// RUnlock it runlocks a specific item by it's ID
-func (s *KeyRWLocker) RUnlock(ID string) {
-	iLocks, _ := s.m.Load(ID)
-	iLock := iLocks.(*sync.RWMutex)
-	iLock.RUnlock()
+// Helper function to check if mutex is locked
+func isLocked(m *sync.Mutex) bool {
```

Comment:
while this method is quite clever it is not safe to use. lock can be acquired anytime between creating variable and returning it (making `locked` value invalid). Even if returned value would be correct, lock still could be acquired between calling `isLocked(...)` method and deleting lock from map.

Reply:
good point